File: Beecrowd_Parte3/InventarioPomekon_2223.c
#include <stdio.h>
#include <stdlib.h>

// Definindo a estrutura da Árvore de Fenwick (BIT)
typedef struct {
    int *tree;
    int size;
} BIT;

// Inicializa a Árvore de Fenwick
BIT *createBIT(int size) {
    BIT *bit = (BIT *)malloc(sizeof(BIT));
    bit->size = size;
    bit->tree = (int *)calloc(size + 1, sizeof(int));
    return bit;
}

// Função para atualizar um elemento na BIT
void update(BIT *bit, int index, int delta) {
    while (index <= bit->size) {
        bit->tree[index] += delta;
        index += index & (-index);
    }
}

// Função para realizar uma consulta de soma até o índice 'index'
int query(BIT *bit, int index) {
    int sum = 0;
    while (index > 0) {
        sum += bit->tree[index];
        index -= index & (-index);
    }
    return sum;
}

// Função para realizar uma consulta de soma no intervalo [left, right]
int queryRange(BIT *bit, int left, int right) {
    return query(bit, right) - query(bit, left - 1);
}

// Libera a memória da BIT
void freeBIT(BIT *bit) {
    free(bit->tree);
    free(bit);
}

int main() {
    int T;
    scanf("%d", &T); // Número de instâncias
    getchar(); // Consumir o newline após T
    
    while (T-- > 0) {
        int N;
        scanf("%d", &N); // Quantidade de pilhas
        int *pilhas = (int *)malloc((N + 1) * sizeof(int));
        
        // Inicialização das pilhas e da BIT
        BIT *bit = createBIT(N);
        for (int i = 1; i <= N; i++) {
            scanf("%d", &pilhas[i]);
            update(bit, i, pilhas[i]);
        }
        
        int Q;
        scanf("%d", &Q); // Quantidade de operações
        
        while (Q-- > 0) {
            int tipoOperacao;
            scanf("%d", &tipoOperacao);
            
            if (tipoOperacao == 1) {
                int X, Y, K;
                scanf("%d %d %d", &X, &Y, &K);
                
                // Consulta quantidade de elementos no intervalo [X, Y]
                int quantidadeXaY = queryRange(bit, X, Y);
                printf("%d\n", quantidadeXaY);
            } else if (tipoOperacao == 2) {
                int X, W;
                scanf("%d %d", &X, &W);
                
                // Atualiza a pilha na posição X com o valor W
                int valorAntigo = pilhas[X];
                int delta = W - valorAntigo;
                pilhas[X] = W;
                update(bit, X, delta);
            }
        }
        
        free(pilhas);
        freeBIT(bit);
    }
    
    return 0;
}

File: Beecrowd_Parte3/InventarioPomekon_2223.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

class BIT {
    int[] tree;
    int size;

    public BIT(int size) {
        this.tree = new int[size + 1];
        this.size = size;
    }

    public void update(int index, int delta) {
        while (index <= size) {
            tree[index] += delta;
            index += index & (-index);
        }
    }

    public int query(int index) {
        int sum = 0;
        while (index > 0) {
            sum += tree[index];
            index -= index & (-index);
        }
        return sum;
    }

    public int queryRange(int left, int right) {
        return query(right) - query(left - 1);
    }
}

class Operacao {
    int indiceOperacao;
    int pilha;
    int valor;

    public Operacao(int indiceOperacao, int pilha, int valor) {
        this.indiceOperacao = indiceOperacao;
        this.pilha = pilha;
        this.valor = valor;
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int T = Integer.parseInt(br.readLine().trim()); // Número de instâncias
        StringBuilder sb = new StringBuilder();

        while (T-- > 0) {
            int N = Integer.parseInt(br.readLine().trim()); // Quantidade de pilhas
            int[] pilhas = new int[N + 1];
            BIT bit = new BIT(N);

            st = new StringTokenizer(br.readLine());
            for (int i = 1; i <= N; i++) {
                pilhas[i] = Integer.parseInt(st.nextToken());
                bit.update(i, pilhas[i]);
            }

            int Q = Integer.parseInt(br.readLine().trim()); // Quantidade de operações
            List<Operacao> operacoesTipo2 = new ArrayList<>();

            for (int i = 1; i <= Q; i++) {
                st = new StringTokenizer(br.readLine());
                int tipoOperacao = Integer.parseInt(st.nextToken());
                if (tipoOperacao == 1) {
                    int X = Integer.parseInt(st.nextToken());
                    int Y = Integer.parseInt(st.nextToken());
                    int K = Integer.parseInt(st.nextToken());

                    int quantidadeXaY = bit.queryRange(X, Y);
                    int indiceOperacao = K;

                    sb.append(quantidadeXaY).append("\n");
                } else if (tipoOperacao == 2) {
                    int X = Integer.parseInt(st.nextToken());
                    int W = Integer.parseInt(st.nextToken());
                    bit.update(X, W);
                    operacoesTipo2.add(new Operacao(i, X, W));
                }
            }
        }

        System.out.print(sb.toString());
    }
}
